presto-jdbc/src/test/java/com/facebook/presto/jdbc/TestProgressMonitor.java

```diff
@@ -0,0 +1,172 @@
+            mapper.writeValue(stringWriter, queryResults);
+            return stringWriter.toString();
+        }
+        catch (Exception e) {
+            throw Throwables.propagate(e);
+        }
+    }
+
+    @Test
+    public void test()
+            throws Exception
+    {
+        try (Connection connection = createConnection()) {
+            try (Statement statement = connection.createStatement()) {
+                PrestoStatement prestoStatement = statement.unwrap(PrestoStatement.class);
+                RecordingProgressMonitor progressMonitor = new RecordingProgressMonitor();
+                prestoStatement.setProgressMonitor(progressMonitor);
+                try (ResultSet rs = statement.executeQuery("SELECT a bogus query")) {
+                    ResultSetMetaData metadata = rs.getMetaData();
+                    assertEquals(metadata.getColumnCount(), 1);
+                    assertEquals(metadata.getColumnName(1), "_col0");
+
+                    assertTrue(rs.next());
+                    assertEquals(rs.getLong(1), 253161L);
+                    assertEquals(rs.getLong("_col0"), 253161L);
+
+                    assertFalse(rs.next());
+                }
+                prestoStatement.clearProgressMonitor();
+
+                List<QueryStats> queryStatsList = progressMonitor.finish();
+                assertGreaterThanOrEqual(queryStatsList.size(), 5); // duplicate stats is possible
+                assertEquals(queryStatsList.get(0).getState(), "QUEUED");
+                assertEquals(queryStatsList.get(queryStatsList.size() - 1).getState(), "FINISHED");
+            }
+        }
+    }
+
+    private class RecordingProgressMonitor
+            implements Consumer<QueryStats>
+    {
+        private boolean finished = false;
+        private final ImmutableList.Builder<QueryStats> builder = ImmutableList.builder();
+
+        @Override
+        public synchronized void accept(QueryStats queryStats)
+        {
+            checkState(!finished);
+            builder.add(queryStats);
+        }
+
+        public synchronized List<QueryStats> finish()
+        {
+            finished = true;
+            return builder.build();
+        }
+    }
+
+    private Connection createConnection()
+            throws Exception
+    {
+        QueryExecutor testQueryExecutor = QueryExecutor.createTestExecutor(
+                "user-agent",
+                new TestingHttpClient(new TestingHttpClientProcessor(RESPONSES)));
+        URI uri = new URI(format("prestotest://%s", SERVER_ADDRESS));
+        return new PrestoConnection(uri, "test", testQueryExecutor);
+    }
+
+    private static class TestingHttpClientProcessor
+            implements TestingHttpClient.Processor
+    {
+        private final List<String> responses;
+        private int index = 0;
+
+        public TestingHttpClientProcessor(List<String> responses)
+        {
+            this.responses = responses;
```

Comment:
Check for null and defensive copy (this is a private class so doesn't really matter, but good practice)